### PR TITLE
allow for tied row attention with uneven number of MSAs per sequence …

### PR DESCRIPTION
**Pull request by lucidrains**
…in a batch

### CODE REVIEW

File: alphafold2_pytorch/alphafold2.py [hunk 1]
```diff
@@ -333,12 +333,15 @@ def forward(self, x, context = None, mask = None, context_mask = None, tie_attn_
         if exists(tie_attn_dim):
             q, k, v = map(lambda t: rearrange(t, '(b r) h n d -> b r h n d', r = tie_attn_dim), (q, k, v))
 
-            # when tying row-attention, one cannot have any masked out tokens
             if exists(mask):
-                assert torch.all(mask), 'you cannot have any padding if you are to tie the row attention across MSAs'
-                mask = None
+                mask = rearrange(mask, '(b r) n -> b r n', r = tie_attn_dim)
+                num_rows = (mask.sum(dim = -1) > 0).sum(dim = -1)
```
Comment (aced125) [inline alphafold2_pytorch/alphafold2.py:338]: Could this line be simplified to: 
```python
num_rows = mask.any(-1).sum(-1)
```

File: alphafold2_pytorch/alphafold2.py [hunk 2]
```diff
@@ -333,12 +333,15 @@ def forward(self, x, context = None, mask = None, context_mask = None, tie_attn_
...
+                num_rows = rearrange(num_rows, 'b -> b () () ()')
+                mask = mask.sum(dim = 1) > 0
```
Comment (aced125) [inline alphafold2_pytorch/alphafold2.py:340]: Similarly over here

```python
mask = mask.any(1)
```

File: alphafold2_pytorch/alphafold2.py [hunk 3]
```diff
@@ -333,12 +333,15 @@ def forward(self, x, context = None, mask = None, context_mask = None, tie_attn_
         if exists(tie_attn_dim):
             q, k, v = map(lambda t: rearrange(t, '(b r) h n d -> b r h n d', r = tie_attn_dim), (q, k, v))
 
-            # when tying row-attention, one cannot have any masked out tokens
             if exists(mask):
-                assert torch.all(mask), 'you cannot have any padding if you are to tie the row attention across MSAs'
-                mask = None
+                mask = rearrange(mask, '(b r) n -> b r n', r = tie_attn_dim)
+                num_rows = (mask.sum(dim = -1) > 0).sum(dim = -1)
+                num_rows = rearrange(num_rows, 'b -> b () () ()')
```
Comment (aced125) [inline alphafold2_pytorch/alphafold2.py:339]: Is this required? I thought it would broadcast automatically - but you might've checked already

**Review by aced125 — APPROVED** [review]
LGTM - could consider using the `.any` method on a Tensor, which checks if there is at least one True along a dimension

**Review by aced125 — CHANGES_REQUESTED** [review]
Actually sorry - I think I've noticed smth new now. I think when doing the einsum, some of the r dimensions need to be masked. See below for more clarity. Running this code snippet results in 100% of the elements being the same, but unhashing the line that simulates a mask when an MSA has too few sequences results in only 80% of the elements being the same.


```python
import torch as th
from einops import rearrange

bsz = 4; num_msa = 3; num_aa = 6; dim = 32; heads = 8
q, k, v = (th.randn(bsz, num_msa, heads, num_aa, dim), ) * 3
mask = th.ones(bsz, num_msa, num_aa, dtype=th.bool)

mask[2, :, 4:] = False  # the MSA in the third batch has only 4 amino acids


# UNHASH this line if you want to simulate what happens when one of the MSAs has too few sequences
# mask[3, -1, :] = False  # the MSA in the fourth batch has only 2 sequences

num_rows = mask.any(-1).sum(-1)

nmask = mask.any(1)
mask_val = -th.finfo(dots.dtype).max
nmask = nmask[:, None, :, None] * nmask[:, None, None, :]  # self attn


##### LUCID IMPLEMENTATION ######
dots = th.einsum("b r h i d, b r h j d -> b h i j", q, k)
attn = dots.masked_fill(~nmask, mask_val).log_softmax(-1)


##### FB IMPLEMENTATION ######
nq = q * rearrange(mask, 'b num_msas seq_len -> b num_msas () seq_len ()')
newdots = th.einsum("b r h i d, b r h j d -> b h i j", nq, k)
newattn = newdots.masked_fill(~nmask, mask_val).log_softmax(-1)


print(f'Both attentions equal? {th.allclose(newattn, attn)}')
print(f' {100 * th.isclose(newattn, attn).sum() / newattn.numel()}% of the elements are the same')
```

**Comment by aced125** [comment]
I think the solution is quite simple - simply add these lines 

```python
if exists(tie_attn_dim):
    q, k, v = map(lambda t: rearrange(t, '(b r) h n d -> b r h n d', r=tie_attn_dim), (q, k, v))
    if exists(mask):
        mask = rearrange(mask, '(b r) n -> b r n', r=tie_attn_dim)

        q *= rearrange(mask, 'b r n -> b r () n ()')  # THE NEW LINE

        num_rows = (mask.sum(dim=-1) > 0).sum(dim=-1)
        num_rows = rearrange(num_rows, 'b -> b () () ()')
        mask = mask.sum(dim=1) > 0
```

**Comment by lucidrains** [comment]
@aced125 got it! thanks for letting me know that any(dim) works (i thought they didn't support axis-wise by their docs)

also fixed the zeroing of the contributions where rows are padding 🙏

**Review by aced125 — APPROVED** [review]
lgtm
